karaf/osgi-jta/src/main/java/org/jboss/narayana/quickstart/osgi/RecoveryExample.java

```diff
@@ -0,0 +1,82 @@
+ * contributors by the @authors tag. See the copyright.txt in the
+ * distribution for a full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.narayana.quickstart.osgi;
+
+import org.apache.karaf.shell.api.action.Action;
+import org.apache.karaf.shell.api.action.Command;
+import org.apache.karaf.shell.api.action.Option;
+import org.apache.karaf.shell.api.action.lifecycle.Reference;
+import org.apache.karaf.shell.api.action.lifecycle.Service;
+import org.jboss.narayana.quickstart.xa.DummyXAResource;
+import org.jboss.narayana.quickstart.xa.DummyXAResourceRecovery;
+import org.jboss.tm.XAResourceRecovery;
+import org.osgi.framework.BundleContext;
+
+import javax.transaction.Transaction;
+import javax.transaction.TransactionManager;
+import java.io.File;
+
+/**
+ * @author <a href="mailto:zfeng@redhat.com">Amos Feng</a>
+ */
+
+@Command(scope = "narayana-quickstart", name = "testRecovery", description="The JTA Recovery Example")
+@Service
+public class RecoveryExample implements Action {
+    @Reference
+    private BundleContext context;
+
+    @Reference
+    private TransactionManager tm;
+
+    @Option(name = "-f", aliases = "--force", description = "force the example to crash", required = false, multiValued = false)
+    boolean force = false;
+
+    @Override
+    public Object execute() throws Exception {
+        if (force) {
+            System.out.println("testRecovery generate something to recovery");
+            if (tm == null) {
+                System.out.println("Can not get transaction manager");
+            } else {
+                DummyXAResource xa1 = new DummyXAResource();
+                DummyXAResource xa2 = new DummyXAResource(DummyXAResource.faultType.HALT);
+
+                tm.begin();
+                Transaction transaction = tm.getTransaction();
+                transaction.enlistResource(xa1);
+                transaction.enlistResource(xa2);
+                tm.commit();
+            }
+        } else {
+            File file = new File("DummyXAResource/");
+            if (file.exists() && file.isDirectory() && file.listFiles().length > 0) {
+                System.out.println("register the DummyXAResourceRecovery");
+                int currentCommits = DummyXAResource.getCommitRequests();
+                context.registerService(XAResourceRecovery.class, new DummyXAResourceRecovery(), null);
```

Comment:
You could look at this to see if it provides clues: https://github.com/FuseByExample/esb-transactions/blob/a694b6f5a8cbbff6598f4e7b79ed00fd764e1c32/routing/src/main/resources/OSGI-INF/blueprint/routing.xml#L28 (search for the word recover)